### DISCUSSION TITLE
Client side routing and client side route protection on SSG Apps

### DISCUSSION THREAD

**Discussion by snake-py** [vikejs/vike#938] [Discussion]
Hello,

### Background 
so I have a new project on which I am going to write a React SPA. My constraint is that I don't want to have any server costs and that is why they want to deploy it statically to an s3. I want to harness the power of this plugin ssg capabilities and pre-render what I can pre-render. I will then also need to have hydration in place, but from what I understand from the docs this plugin got me covered. 

### Question
In the past I have used react-router-dom for such projects. They offer me the feature that I can simply protect routes and hook into the router functionality. So I am talking about "client side protection" I know this is no "real" protection, but is good enough since all the data is protected server side and the rendered components do not expose anything of value. 

I want a functionality inside the router which can perform a custom check, e.g. call an external api to see if a token is still valid (maybe a default check if the token is still there), on success then the router would proceed to the correct page on fail the router would redirect to login (or an other defined page). Awesome would be if then on login success I could redirect to the initial desired route

Example:

```
page A (guarded)
page B (guarded)
page C (unguarded login page)
```

 Unauthenticated User comes and 
- visits page A
- router redirects to page C
- user logs in 
- router redirects automatically to page A

Authenticated User comes and visits page C
- router detects that this route is only available for guests and redirects to authenticated home

**Comment by brillout** [comment]
Usually, such logic is implemented before calling [`renderPage()`](https://vite-plugin-ssr.com/renderPage).

> My constraint is that I don't want to have any server costs and that is why they want to deploy it statically

In that case `renderPage()` isn't a solution. Instead call your authentication API in the [client-side render hook](https://vite-plugin-ssr.com/render-client), or call it in an async Route Function. (FYI async Route Functions will soon be deprecated in favor of [Route Guards](https://github.com/brillout/vite-plugin-ssr/issues/939).) Then perform a [redirection on the client-side](https://vite-plugin-ssr.com/page-redirection#client-side) if the user isn't authorized.

> Awesome would be if then on login success I could redirect to the initial desired route

I'd recommend this then: https://vite-plugin-ssr.com/page-redirection#authentication.

**Reply by snake-py** [reply]
@brillout I am tomorrow in holiday, but I will try this path after I come back. We'll see if this works. I am unsure if I fully understand everything. Also I am no friend of file based routing tbh. I rather like having everything defined by code. I saw that this is also fully supported :) Excited to use this package.

**Reply by brillout** [reply]
> Also I am no friend of file based routing tbh. I rather like having everything defined by code.

It's on the roadmap, see https://github.com/brillout/vite-plugin-ssr/issues/341.

**Reply by snake-py** [reply]
So I am playing now around with this and I am a little confused, do I even need the server file in case I only want to do SSG in my SPA?

**Reply by brillout** [reply]
Depends on what you want/need.
